### PR TITLE
Fix style issues related to NumPy

### PR DESCRIPTION
**Pull request by shwina**
NumPy 1.20 is [typed](https://numpy.org/devdocs/release/1.20.0-notes.html#numpy-is-now-typed), which exposed a few typing errors in cuDF that this PR addresses.

### CODE REVIEW

File: python/cudf/cudf/core/column/categorical.py
Comment (kkraus14) [inline python/cudf/cudf/core/column/categorical.py:133]: Why is this `Optional`? Doesn't this always return True or False?

File: python/cudf/cudf/core/column/numerical.py
Comment (kkraus14) [inline python/cudf/cudf/core/column/numerical.py:481]: Any other options here? I know this is used for typing purposes, but maybe we should handle `None` more explicitly somehow or something else?

File: python/cudf/cudf/core/column/datetime.py
Comment (kkraus14) [inline python/cudf/cudf/core/column/datetime.py:258]: Does this work for older versions of Numpy as well?

File: python/cudf/cudf/core/column/categorical.py
Comment (shwina) [reply python/cudf/cudf/core/column/categorical.py:133]: `None` (at least in Pandas) has different semantics from `False` (https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.CategoricalDtype.html).

I'm not sure if we actually support it, but currently in cuDF, we do allow `ordered=None`:

```
In [15]: import cudf

In [16]: dt = cudf.CategoricalDtype(categories=[1, 2, 3])

In [17]: dt.ordered is None
Out[17]: True
```

File: python/cudf/cudf/core/column/numerical.py
Comment (shwina) [reply python/cudf/cudf/core/column/numerical.py]: Yup - raising a `ValueError` instead both gets rid of the mypy error and makes the code better. FIxed.

File: python/cudf/cudf/core/column/numerical.py
Comment (kkraus14) [inline python/cudf/cudf/core/column/numerical.py:475]: Do we actually not allow `None` here or does that mean fill with `null` values?

File: python/cudf/cudf/core/column/datetime.py
Comment (shwina) [reply python/cudf/cudf/core/column/datetime.py:258]: Yes:

```
In [3]: np.dtype(np.bool) == np.dtype(np.bool_)
Out[3]: True

In [4]: np.__version__
Out[4]: '1.19.4'
```

`np.bool` has disappeared from numpy, so we should use either just `bool` or `np.bool_/np.bool8` going forward.

File: python/cudf/cudf/core/column/numerical.py
Comment (shwina) [reply python/cudf/cudf/core/column/numerical.py:475]: `None` is allowed when you use `method`:

```
In [8]: s = cudf.Series([1, 2, None, None, 3])

In [9]: s.fillna(method='ffill')
Out[9]:
0    1
1    2
2    2
3    2
4    3
dtype: int64
```

Comment (shwina) [reply python/cudf/cudf/core/column/numerical.py:475]: Sorry - I misunderstood the question. We should explicitly _not_ allow `None` when `method=None`. 

If someone wants to fillna() with null values, they should just do nothing instead :)

Comment (kkraus14) [reply python/cudf/cudf/core/column/numerical.py:475]: > If someone wants to fillna() with null values, they should just do nothing instead :)

This isn't true in the case of `float` dtype columns though, no? Since we handle both `NaN` and `null` values via `fillna` in that case.

Comment (shwina) [reply python/cudf/cudf/core/column/numerical.py:475]: What case are we thinking, where someone wants to convert `nans` to nulls? Or nulls to nans?

Comment (shwina) [reply python/cudf/cudf/core/column/numerical.py:475]: In the former case, they could use [`nans_to_nulls()`](https://github.com/rapidsai/cudf/blob/branch-0.18/python/cudf/cudf/core/series.py#L1960).

In the latter case, they could use `fillna(np.nan)` which is much more explicit.

FWIW, we currently throw when both `method=None` and `fill_value=None` further down: https://github.com/rapidsai/cudf/blob/branch-0.18/python/cudf/cudf/_lib/replace.pyx#L134

File: python/cudf/cudf/core/column/datetime.py
Comment (kkraus14) [reply python/cudf/cudf/core/column/datetime.py:258]: I meant the `# type: Dtype` 😄

Comment (shwina) [reply python/cudf/cudf/core/column/datetime.py:258]: That's our own type :)

**Review by kkraus14 — APPROVED** [review]

**Comment by kkraus14** [comment]
@gpucibot merge

File: python/cudf/cudf/_lib/column.pyi
Comment (galipremsagar) [inline python/cudf/cudf/_lib/column.pyi:27]: was this an intentional change(un-typing)?

Comment (shwina) [reply python/cudf/cudf/_lib/column.pyi]: Nope - unintentional. Just fixed - thanks!

**Review by galipremsagar — APPROVED** [review]

File: python/cudf/cudf/core/column/column.py
Comment (kkraus14) [inline python/cudf/cudf/core/column/column.py:1515]: We shouldn't rely on assertions, how can we be more explicit?

Comment (shwina) [reply python/cudf/cudf/core/column/column.py]: Replaced the assertions with TypeErrors, add TODOs to make the `size` param optional in column constructors.

**Comment by kkraus14** [comment]
@gpucibot merge

**Comment by harrism** [comment]
CI failure: https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/cudf/job/prb/job/cudf-gpu-test/CUDA=10.1,GPU_LABEL=gpu,OS=ubuntu16.04,PYTHON=3.7/623/testReport/cudf.tests/test_dataframe/test_series_from_cupy_scalars/

File: ci/gpu/build.sh
Comment (ajschmidt8) [inline ci/gpu/build.sh:86]: this will need to be removed before the PR is merged. requesting changes here to prevent @kkraus14's `@gpucibot merge` comment from automatically merging this PR.

**Review by ajschmidt8 — CHANGES_REQUESTED** [review]

File: ci/gpu/build.sh
Comment (kkraus14) [reply ci/gpu/build.sh:86]: @ajschmidt8 this is temporary to unblock CI as there's a bug in upstream Dask DataFrame related to numpy 1.20. We plan to fix that upstream and then remove this in a future PR.

Other libraries like cuML really want numpy 1.20 as it fixes other bugs for them so we don't want to pin `integration`.

**Review by ajschmidt8 — APPROVED** [review]

**Pull request merged**
